src/lib/block/aes/aes_vaes/aes_vaes.cpp

```diff
@@ -0,0 +1,625 @@
+   const SIMD_8x32 K2 = SIMD_8x32::load_le128(&m_EK[4 * 2]);
+   const SIMD_8x32 K3 = SIMD_8x32::load_le128(&m_EK[4 * 3]);
+   const SIMD_8x32 K4 = SIMD_8x32::load_le128(&m_EK[4 * 4]);
+   const SIMD_8x32 K5 = SIMD_8x32::load_le128(&m_EK[4 * 5]);
+   const SIMD_8x32 K6 = SIMD_8x32::load_le128(&m_EK[4 * 6]);
+   const SIMD_8x32 K7 = SIMD_8x32::load_le128(&m_EK[4 * 7]);
+   const SIMD_8x32 K8 = SIMD_8x32::load_le128(&m_EK[4 * 8]);
+   const SIMD_8x32 K9 = SIMD_8x32::load_le128(&m_EK[4 * 9]);
+   const SIMD_8x32 K10 = SIMD_8x32::load_le128(&m_EK[4 * 10]);
+
+   while(blocks >= 8) {
+      SIMD_8x32 B0 = SIMD_8x32::load_le(in);
+      SIMD_8x32 B1 = SIMD_8x32::load_le(in + 16 * 2);
+      SIMD_8x32 B2 = SIMD_8x32::load_le(in + 16 * 4);
+      SIMD_8x32 B3 = SIMD_8x32::load_le(in + 16 * 6);
+
+      keyxor(K0, B0, B1, B2, B3);
+      aesenc(K1, B0, B1, B2, B3);
+      aesenc(K2, B0, B1, B2, B3);
+      aesenc(K3, B0, B1, B2, B3);
+      aesenc(K4, B0, B1, B2, B3);
+      aesenc(K5, B0, B1, B2, B3);
+      aesenc(K6, B0, B1, B2, B3);
+      aesenc(K7, B0, B1, B2, B3);
+      aesenc(K8, B0, B1, B2, B3);
+      aesenc(K9, B0, B1, B2, B3);
+      aesenclast(K10, B0, B1, B2, B3);
+
+      B0.store_le(out);
+      B1.store_le(out + 16 * 2);
+      B2.store_le(out + 16 * 4);
+      B3.store_le(out + 16 * 6);
+
+      blocks -= 8;
+      in += 8 * 16;
+      out += 8 * 16;
+   }
+
+   while(blocks >= 2) {
+      SIMD_8x32 B = SIMD_8x32::load_le(in);
+
+      B ^= K0;
+      aesenc(K1, B);
+      aesenc(K2, B);
+      aesenc(K3, B);
+      aesenc(K4, B);
+      aesenc(K5, B);
+      aesenc(K6, B);
+      aesenc(K7, B);
+      aesenc(K8, B);
+      aesenc(K9, B);
+      aesenclast(K10, B);
+
+      B.store_le(out);
+
+      in += 2 * 16;
+      out += 2 * 16;
+      blocks -= 2;
+   }
+
+   if(blocks > 0) {
+      SIMD_8x32 B = SIMD_8x32::load_le128(in);
+
+      B ^= K0;
+      aesenc(K1, B);
+      aesenc(K2, B);
+      aesenc(K3, B);
+      aesenc(K4, B);
+      aesenc(K5, B);
+      aesenc(K6, B);
+      aesenc(K7, B);
+      aesenc(K8, B);
+      aesenc(K9, B);
+      aesenclast(K10, B);
+
+      B.store_le128(out);
+   }
+}
+
+/*
+* AES-128 Decryption
+*/
+BOTAN_FUNC_ISA("vaes,avx2") void AES_128::x86_vaes_decrypt_n(const uint8_t in[], uint8_t out[], size_t blocks) const {
+   const SIMD_8x32 K0 = SIMD_8x32::load_le128(&m_DK[4 * 0]);
+   const SIMD_8x32 K1 = SIMD_8x32::load_le128(&m_DK[4 * 1]);
+   const SIMD_8x32 K2 = SIMD_8x32::load_le128(&m_DK[4 * 2]);
+   const SIMD_8x32 K3 = SIMD_8x32::load_le128(&m_DK[4 * 3]);
+   const SIMD_8x32 K4 = SIMD_8x32::load_le128(&m_DK[4 * 4]);
+   const SIMD_8x32 K5 = SIMD_8x32::load_le128(&m_DK[4 * 5]);
+   const SIMD_8x32 K6 = SIMD_8x32::load_le128(&m_DK[4 * 6]);
+   const SIMD_8x32 K7 = SIMD_8x32::load_le128(&m_DK[4 * 7]);
+   const SIMD_8x32 K8 = SIMD_8x32::load_le128(&m_DK[4 * 8]);
+   const SIMD_8x32 K9 = SIMD_8x32::load_le128(&m_DK[4 * 9]);
+   const SIMD_8x32 K10 = SIMD_8x32::load_le128(&m_DK[4 * 10]);
```

Comment:
I kind of have an itch to evaluate the possibility of `std::array<SIMD_8x32, 11> K = load_le<SIMD_8x32>(...)` (and friends). Shouldn't be too complicated to pull that off, I think.

Reply:
This particular case is complicated because we're doing a 128 bit load then broadcasting to the high part. It's not clear if this is something generally useful, anyway it hasn't come up before now.